File: python/examples/ex06_webhook_server.py
"""Example 06 -- Webhook receiver with HMAC verification (stdlib http.server only).

Purpose
    Receive Coasty run lifecycle webhooks, verify the ``Coasty-Signature``
    header via :func:`coasty.webhooks.verify_signature` (HMAC-SHA256 over
    ``"<t>." + raw_body``, constant-time compare, +/- 5 minute timestamp
    tolerance), and dispatch per-event handlers.

Flow
    Coasty POSTs JSON  ->  verify signature (401 on invalid / stale /
    tampered / malformed)  ->  parse the signed body  ->  dispatch on
    ``payload["event"]``  ->  respond 200 FAST. Handlers here only log;
    anything slow (DB writes, resuming runs) belongs on a queue / thread so
    the 200 goes back immediately and Coasty never retries spuriously.

Endpoints
    None called -- this is the *receiving* side. Runs created with a
    ``webhook_url`` get callbacks for run.succeeded / run.failed /
    run.cancelled / run.timed_out / run.awaiting_human, each signed with the
    run's one-time ``webhook_secret`` (returned exactly once on create).

Estimated cost
    $0.00 -- receiving webhooks is free (coasty.cost: no billable operation
    in this example), so there is no spend gate.

Run
    COASTY_WEBHOOK_SECRET=whsec_... python examples/ex06_webhook_server.py \
        --host 127.0.0.1 --port 8788
"""

from __future__ import annotations

import argparse
import json
import logging
import os
import sys
from collections.abc import Callable, Mapping, Sequence
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any

from coasty import verify_signature

logger = logging.getLogger("coasty.examples.ex06")

SIGNATURE_HEADER = "Coasty-Signature"
SECRET_ENV_VAR = "COASTY_WEBHOOK_SECRET"

EventHandler = Callable[[dict[str, Any]], str]


# ── per-event handlers (keep them FAST; offload slow work) ─────────────────


def _run_id_of(payload: Mapping[str, Any]) -> str:
    run_id = payload.get("run_id")
    return run_id if isinstance(run_id, str) else "<unknown>"


def on_run_succeeded(payload: dict[str, Any]) -> str:
    action = f"acknowledged run.succeeded for run {_run_id_of(payload)}"
    logger.info("%s", action)
    return action


def on_run_failed(payload: dict[str, Any]) -> str:
    action = f"acknowledged run.failed for run {_run_id_of(payload)} -- alert the on-call"
    logger.warning("%s", action)
    return action


def on_run_cancelled(payload: dict[str, Any]) -> str:
    action = f"acknowledged run.cancelled for run {_run_id_of(payload)}"
    logger.info("%s", action)
    return action


def on_run_timed_out(payload: dict[str, Any]) -> str:
    action = f"acknowledged run.timed_out for run {_run_id_of(payload)} -- consider a retry"
    logger.warning("%s", action)
    return action


def on_run_awaiting_human(payload: dict[str, Any]) -> str:
    """The run paused for a human (captcha, 2FA, ambiguous UI, ...).

    This is exactly where a real operator console would, once a human has
    cleared the blocker, hand control back to the agent:

        from coasty import CoastyClient

        with CoastyClient() as client:
            result = client.resume_run(run_id, note="human solved the captcha")

    ``POST /v1/runs/{id}/resume`` is only valid while the run is in
    ``awaiting_human`` -- anything else returns ``409 NOT_AWAITING_HUMAN``.
    Do NOT call it synchronously from this handler: respond 200 first, then
    resume from a worker once the human is actually done.
    """
    run_id = _run_id_of(payload)
    reason = payload.get("reason")
    action = (
        f"resume_required: run {run_id} is awaiting a human (reason={reason!r}) -- "
        "call runs.resume once resolved"
    )
    logger.warning("%s", action)
    return action


EVENT_HANDLERS: dict[str, EventHandler] = {
    "run.succeeded": on_run_succeeded,
    "run.failed": on_run_failed,
    "run.cancelled": on_run_cancelled,
    "run.timed_out": on_run_timed_out,
    "run.awaiting_human": on_run_awaiting_human,
}


# ── pure, socket-free core (this is what the unit tests drive) ─────────────


def handle_webhook(
    raw_body: bytes,
    headers: Mapping[str, str],
    secret: str,
    *,
    now: int | None = None,
) -> tuple[int, str]:
    """Verify + dispatch one webhook delivery; returns ``(status, action)``.

    - 401 when the ``Coasty-Signature`` header is missing, malformed, signed
      with the wrong secret, computed over a tampered body, or its timestamp
      is outside the +/- 5 minute tolerance (``now`` is injectable for tests).
    - 400 when the signature is valid but the body is not a JSON object.
    - 200 otherwise; unknown event types are acknowledged and ignored.
    """
    header = next(
        (value for key, value in headers.items() if key.lower() == SIGNATURE_HEADER.lower()),
        None,
    )
    if header is None:
        return 401, f"rejected: missing {SIGNATURE_HEADER} header"
    if not verify_signature(raw_body, header, secret, now=now):
        return 401, "rejected: invalid, stale, or tampered signature"

    try:
        payload = json.loads(raw_body)
    except (json.JSONDecodeError, UnicodeDecodeError):
        return 400, "rejected: signed body is not valid JSON"
    if not isinstance(payload, dict):
        return 400, "rejected: signed body is not a JSON object"

    event = payload.get("event")
    handler = EVENT_HANDLERS.get(event) if isinstance(event, str) else None
    if handler is None:
        return 200, f"ignored: unhandled event {event!r}"
    return 200, handler(payload)


# ── thin stdlib HTTP wrapper around the core ───────────────────────────────


class WebhookRequestHandler(BaseHTTPRequestHandler):
    """Reads the raw body, delegates to :func:`handle_webhook`, replies fast."""

    secret: str = ""  # bound per-server by make_server()

    def do_POST(self) -> None:  # stdlib handler hook (name fixed by http.server)
        length_header = self.headers.get("Content-Length", "0")
        try:
            length = int(length_header)
        except ValueError:
            length = 0
        raw_body = self.rfile.read(length) if length > 0 else b""

        status, action = handle_webhook(raw_body, dict(self.headers.items()), self.secret)

        body = json.dumps({"received": status == 200, "action": action}).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, format: str, *args: Any) -> None:  # stdlib hook signature
        logger.info("%s -- %s", self.address_string(), format % args)


def make_server(host: str, port: int, secret: str) -> ThreadingHTTPServer:
    """Build a webhook server bound to ``secret`` (port 0 = ephemeral)."""

    class BoundHandler(WebhookRequestHandler):
        pass

    BoundHandler.secret = secret
    return ThreadingHTTPServer((host, port), BoundHandler)


# ── CLI ────────────────────────────────────────────────────────────────────


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    parser.add_argument("--host", default="127.0.0.1", help="bind address (default 127.0.0.1)")
    parser.add_argument("--port", type=int, default=8788, help="bind port (default 8788)")
    args = parser.parse_args(argv)

    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")

    secret = os.environ.get(SECRET_ENV_VAR, "").strip()
    if not secret:
        print(
            f"error: {SECRET_ENV_VAR} is not set. Use the one-time webhook_secret "
            "returned when the run was created (never logged here).",
            file=sys.stderr,
        )
        return 2

    server = make_server(args.host, args.port, secret)
    raw_host, port = server.server_address[0], server.server_address[1]
    host = raw_host.decode() if isinstance(raw_host, bytes) else raw_host
    print(f"listening on http://{host}:{port} -- POST signed webhooks here (Ctrl+C to stop)")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("shutting down")
    finally:
        server.server_close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/src/examples/ex06-webhook-server.ts
/**
 * Example 06 — Webhook receiver (node:http stdlib only).
 *
 * Coasty POSTs a signed JSON payload to your `webhook_url` on run lifecycle
 * transitions: `run.succeeded`, `run.failed`, `run.cancelled`,
 * `run.timed_out`, and `run.awaiting_human`. Every callback carries
 * `Coasty-Signature: t=<unix>,v1=<hex>` where the signed payload is
 * `"<t>." + raw_body` and `v1 = hex(HMAC_SHA256(webhook_secret, payload))`.
 *
 * This receiver:
 *   - verifies the signature with a CONSTANT-TIME compare and a ±5 minute
 *     timestamp tolerance (via the shared `webhooks.ts` verifier);
 *   - answers 401 for invalid, stale, or tampered deliveries;
 *   - answers 200 FAST (dispatch happens after the response is written, so a
 *     slow handler can never trigger sender-side retries);
 *   - on `run.awaiting_human` shows exactly where `client.runs.resume(...)`
 *     would be called to hand control back to the agent.
 *
 * Estimated cost: 0 credits ($0.00) — this is a receiver; it makes no API
 * calls. (Resuming a paused run is also free; only completed run STEPS bill,
 * at 5 credits / $0.05 each on v3/v4.)
 *
 * Run it:
 *   COASTY_WEBHOOK_SECRET=whsec_... npx tsx src/examples/ex06-webhook-server.ts [--port 8788]
 *
 * The secret comes from the `webhook_secret` field returned ONCE by
 * `POST /v1/runs` when you pass a `webhook_url` — store it then, export it as
 * `COASTY_WEBHOOK_SECRET` here. Never commit it.
 */
import { createServer, type IncomingMessage, type Server, type ServerResponse } from 'node:http';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { loadEnvFile } from '../coasty/env.js';
import { verifySignature } from '../coasty/webhooks.js';

// ---------------------------------------------------------------------------
// Pure core — fully testable without sockets
// ---------------------------------------------------------------------------

/** The five documented run webhook events. */
export const RUN_WEBHOOK_EVENTS = [
  'run.succeeded',
  'run.failed',
  'run.cancelled',
  'run.timed_out',
  'run.awaiting_human',
] as const;

export type RunWebhookEventName = (typeof RUN_WEBHOOK_EVENTS)[number];

/** A verified, well-formed run webhook payload. */
export interface RunWebhookEvent {
  event: RunWebhookEventName;
  run_id: string;
  /** Terminal events carry the run status (e.g. "succeeded"). */
  status?: string;
  /** `run.awaiting_human` carries the pause reason (e.g. "captcha"). */
  reason?: string;
  /** Any extra documented fields ride along untouched. */
  extra: Record<string, unknown>;
}

export type WebhookDisposition =
  | 'ok'
  | 'ignored_unknown_event'
  | 'invalid_signature'
  | 'invalid_payload';

export interface WebhookOutcome {
  /** HTTP status the receiver should answer with (200 | 400 | 401). */
  status: 200 | 400 | 401;
  disposition: WebhookDisposition;
  /** Present only when the delivery verified AND parsed as a known event. */
  event?: RunWebhookEvent;
}

function parseEvent(rawBody: string): RunWebhookEvent | 'invalid_payload' | 'unknown_event' {
  let parsed: unknown;
  try {
    parsed = JSON.parse(rawBody);
  } catch {
    return 'invalid_payload';
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    return 'invalid_payload';
  }
  const record = parsed as Record<string, unknown>;
  const { event, run_id: runId, status, reason, ...extra } = record;
  if (typeof event !== 'string' || typeof runId !== 'string' || runId === '') {
    return 'invalid_payload';
  }
  if (!(RUN_WEBHOOK_EVENTS as readonly string[]).includes(event)) {
    return 'unknown_event';
  }
  const result: RunWebhookEvent = { event: event as RunWebhookEventName, run_id: runId, extra };
  if (typeof status === 'string') result.status = status;
  if (typeof reason === 'string') result.reason = reason;
  return result;
}

/**
 * Pure webhook handler: verify the `Coasty-Signature` header over the EXACT
 * raw body bytes, then parse + classify the event.
 *
 *   - bad/missing/stale/tampered signature -> 401 (never reveal which check failed)
 *   - verified but not a JSON object with `event` + `run_id` -> 400
 *   - verified, well-formed, unknown event name -> 200 (ack so the sender
 *     does not retry forever) without dispatching
 *   - verified known event -> 200 + the event
 *
 * @param rawBody The exact request bytes (NOT re-serialized JSON).
 * @param header  The `Coasty-Signature` header value, or undefined if absent.
 * @param secret  The per-run `webhook_secret` (shown once on run create).
 * @param now     Unix seconds "now" (injectable for tests; defaults to wall clock).
 */
export function handleWebhook(
  rawBody: Uint8Array | string,
  header: string | undefined,
  secret: string,
  now: number = Math.floor(Date.now() / 1000),
): WebhookOutcome {
  if (header === undefined || !verifySignature(rawBody, header, secret, { now })) {
    return { status: 401, disposition: 'invalid_signature' };
  }
  const text = typeof rawBody === 'string' ? rawBody : Buffer.from(rawBody).toString('utf8');
  const parsed = parseEvent(text);
  if (parsed === 'invalid_payload') return { status: 400, disposition: 'invalid_payload' };
  if (parsed === 'unknown_event') return { status: 200, disposition: 'ignored_unknown_event' };
  return { status: 200, disposition: 'ok', event: parsed };
}

/**
 * What this receiver does for each verified event. Returns the log line so
 * tests can assert the dispatch table — including that `run.awaiting_human`
 * points at `client.runs.resume(...)`.
 */
export function dispatchWebhookEvent(event: RunWebhookEvent): string {
  switch (event.event) {
    case 'run.succeeded':
      return `[${event.run_id}] succeeded — fetch the result via client.runs.get('${event.run_id}')`;
    case 'run.failed':
      return `[${event.run_id}] failed — inspect run.error via client.runs.get('${event.run_id}')`;
    case 'run.cancelled':
      return `[${event.run_id}] cancelled — terminal state, nothing to do`;
    case 'run.timed_out':
      return `[${event.run_id}] timed out — consider a higher deadline_seconds and retry`;
    case 'run.awaiting_human':
      // This is where a real operator UI would page someone. Once the human
      // has resolved the blocker (e.g. a captcha), hand control back with:
      //
      //   await client.runs.resume(event.run_id, { note: 'captcha solved' });
      //
      // (resume is only valid from awaiting_human; otherwise the API answers
      // 409 NOT_AWAITING_HUMAN).
      return (
        `[${event.run_id}] PAUSED (${event.reason ?? 'no reason given'}) — resolve it, then call ` +
        `client.runs.resume('${event.run_id}', { note: '...' }) to continue`
      );
  }
}

// ---------------------------------------------------------------------------
// node:http server (stdlib only)
// ---------------------------------------------------------------------------

/** Deliveries larger than this are rejected outright (Coasty payloads are tiny). */
export const MAX_BODY_BYTES = 1_048_576;

export interface WebhookServerOptions {
  secret: string;
  /** Unix-seconds clock, injectable for tests. */
  now?: () => number;
  /** Called AFTER the 200 has been written, once per verified known event. */
  onEvent?: (event: RunWebhookEvent) => void;
  logger?: (line: string) => void;
}

function readRawBody(request: IncomingMessage, maxBytes: number): Promise<Buffer | null> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    let total = 0;
    request.on('data', (chunk: Buffer) => {
      total += chunk.length;
      if (total > maxBytes) {
        resolve(null); // too large
        request.destroy();
        return;
      }
      chunks.push(chunk);
    });
    request.on('end', () => {
      resolve(Buffer.concat(chunks));
    });
    request.on('error', reject);
  });
}

/**
 * Build (but do not bind) the webhook receiver. The caller picks the port —
 * tests listen on 127.0.0.1:0 (an ephemeral loopback port).
 */
export function createWebhookServer(options: WebhookServerOptions): Server {
  const log = options.logger ?? ((line: string): void => void process.stdout.write(`${line}\n`));
  const now = options.now ?? ((): number => Math.floor(Date.now() / 1000));

  const respond = (response: ServerResponse, status: number, body: string): void => {
    response.writeHead(status, { 'content-type': 'application/json' });
    response.end(JSON.stringify({ status: body }));
  };

  return createServer((request, response) => {
    void (async (): Promise<void> => {
      if (request.method !== 'POST') {
        respond(response, 405, 'method_not_allowed');
        return;
      }
      const rawBody = await readRawBody(request, MAX_BODY_BYTES);
      if (rawBody === null) {
        respond(response, 413, 'payload_too_large');
        return;
      }
      const header = request.headers['coasty-signature'];
      const headerValue = Array.isArray(header) ? header[0] : header;
      const outcome = handleWebhook(rawBody, headerValue, options.secret, now());

      // Answer FIRST (200 fast — the sender's retry clock stops here), then
      // dispatch on the next tick so slow handlers never delay the ack.
      respond(response, outcome.status, outcome.disposition);

      const { event } = outcome;
      if (event !== undefined) {
        setImmediate(() => {
          log(dispatchWebhookEvent(event));
          options.onEvent?.(event);
        });
      } else if (outcome.status !== 200) {
        log(`rejected delivery: ${outcome.disposition} (answered ${String(outcome.status)})`);
      }
    })().catch((error: unknown) => {
      // Never leak internals to the sender; log and fail closed.
      console.error(`webhook receiver error: ${String(error)}`);
      if (!response.headersSent) respond(response, 500, 'internal_error');
    });
  });
}

// ---------------------------------------------------------------------------
// Thin CLI
// ---------------------------------------------------------------------------

function isMain(): boolean {
  const entry = process.argv[1];
  return entry !== undefined && path.resolve(entry) === fileURLToPath(import.meta.url);
}

export function main(argv: string[] = process.argv.slice(2)): void {
  loadEnvFile();
  const secret = process.env.COASTY_WEBHOOK_SECRET?.trim();
  if (secret === undefined || secret === '') {
    console.error(
      'COASTY_WEBHOOK_SECRET is not set. It is the webhook_secret returned ONCE by ' +
        'POST /v1/runs when you pass a webhook_url — export it before starting the receiver.',
    );
    process.exitCode = 1;
    return;
  }

  const portFlag = argv.indexOf('--port');
  const portRaw = portFlag !== -1 ? argv[portFlag + 1] : process.env.COASTY_WEBHOOK_PORT;
  const port = portRaw === undefined ? 8788 : Number(portRaw);
  if (!Number.isInteger(port) || port < 0 || port > 65535) {
    console.error(`invalid port: ${String(portRaw)}`);
    process.exitCode = 1;
    return;
  }

  const print = (line: string): void => void process.stdout.write(`${line}\n`);
  const server = createWebhookServer({ secret, logger: print });
  server.listen(port, '127.0.0.1', () => {
    print(`Coasty webhook receiver listening on http://127.0.0.1:${String(port)}`);
    print('Estimated cost: 0 credits ($0.00) — receiver only, no API calls.');
    print('Point your run webhook_url here (via a tunnel — Coasty requires https).');
  });
}

if (isMain()) main();
